### PR TITLE
refactor: enable -Wconversion warning for mbyte.c

### PR DESCRIPTION
**Pull request by dundargoc**
Work on https://github.com/neovim/neovim/issues/567

### CODE REVIEW

File: src/nvim/mbyte.c [hunk 1]
```diff
@@ -991,40 +991,41 @@ int utf_char2len(const int c)
 /// @param c character to convert to \p buf
 /// @param[out] buf UTF-8 string generated from \p c, does not add \0
 /// @return Number of bytes (1-6).
-int utf_char2bytes(const int c, char *const buf)
+int utf_char2bytes(const int c, char *const buf_in)
 {
+  unsigned *buf = (unsigned *)buf_in;
```
Comment (zeertzjq) [inline src/nvim/mbyte.c:996]: This is not going to work.

Comment (dundargoc) [reply src/nvim/mbyte.c:996]: Eh, you're right. Thanks for the heads up.

File: src/nvim/mbyte.c [hunk 2]
```diff
@@ -994,37 +994,37 @@ int utf_char2len(const int c)
...
 {
   if (c < 0x80) {  // 7 bits
-    buf[0] = c;
+    buf[0] = (char)c;
     return 1;
   } else if (c < 0x800) {  // 11 bits
-    buf[0] = 0xc0 + ((unsigned)c >> 6);
-    buf[1] = 0x80 + (c & 0x3f);
+    buf[0] = (char)(0xc0 + ((unsigned)c >> 6));
+    buf[1] = (char)(0x80 + (unsigned)(c & 0x3f));
```
Comment (philix) [inline src/nvim/mbyte.c:1001]: I think it's better to convert `c` to `unsigned` before the bitwise operation line done in the line above this one.

File: src/nvim/mbyte.c [hunk 3]
```diff
@@ -994,37 +994,37 @@ int utf_char2len(const int c)
...
     return 2;
   } else if (c < 0x10000) {  // 16 bits
-    buf[0] = 0xe0 + ((unsigned)c >> 12);
-    buf[1] = 0x80 + (((unsigned)c >> 6) & 0x3f);
-    buf[2] = 0x80 + (c & 0x3f);
+    buf[0] = (char)(0xe0 + ((unsigned)c >> 12));
+    buf[1] = (char)(0x80 + (((unsigned)c >> 6) & 0x3f));
+    buf[2] = (char)(0x80 + (unsigned)(c & 0x3f));
```
Comment (philix) [inline src/nvim/mbyte.c:1006]: same here

Comment (philix) [reply src/nvim/mbyte.c:1006]: unless you have a specific reason to do it like this

File: src/nvim/mbyte.c [hunk 4]
```diff
@@ -2165,7 +2165,7 @@ char_u *enc_canonize(char_u *enc) FUNC_ATTR_NONNULL_RET
     if (*s == '_') {
       *p++ = '-';
     } else {
-      *p++ = TOLOWER_ASC(*s);
+      *p++ = (uint8_t)TOLOWER_ASC(*s);
```
Comment (philix) [inline src/nvim/mbyte.c:2168]: Could use `char_u` for consistency with the rest of the function.

File: src/nvim/mbyte.c [hunk 5]
```diff
@@ -2550,10 +2550,10 @@ char_u *string_convert_ext(const vimconv_T *const vcp, char_u *ptr, size_t *lenp
     for (size_t i = 0; i < len; ++i) {
       c = ptr[i];
       if (c < 0x80) {
-        *d++ = c;
+        *d++ = (uint8_t)c;
       } else {
-        *d++ = 0xc0 + ((unsigned)c >> 6);
-        *d++ = 0x80 + (c & 0x3f);
+        *d++ = (uint8_t)(0xc0 + (uint8_t)((unsigned)c >> 6));
+        *d++ = (uint8_t)(0x80 + (c & 0x3f));
```
Comment (philix) [inline src/nvim/mbyte.c:2556]: `char_u` for consistency/ease-of-replacement

**Comment by philix** [comment]
LGTM

(It's been years since I looked at the Neovim PRs. Felt good to give a drive-by review, but I will let others give the final approval.)

**Pull request merged**